### ISSUE TITLE
reportGeneralTypeIssues when overwriting type with default value, when removing optional.

### ISSUE THREAD

**Issue by ViktorClaesson** [microsoft/pyright#9487]
I want to pre-face that I understand that this might be unique to my usecase and this might not be applicable for everyone.
However with how Pydantic works I think it is reasonable that removing the default value when reducing the scope of the type to be reasonable.

## Environment data

-   Pylance version: 2024.11.2 (I think, I'm using the latest that comes with VSCode)
-   macOS and version: 14.7.1 (23H222)
-   Python version: 3.12.7
-   Pydantic: 2.9.2
-   Pydantic-core: 2.23.4

## Code Snippet

```python
from pydantic import BaseModel

class A(BaseModel):
    a: int | None = None

class B(A):
    a: int
```

Gives error:
```
"a" overrides a field of the same name but is missing a default valuePylance[reportGeneralTypeIssues](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportGeneralTypeIssues)
```

Which is not correct because when overwriting the type the default value is overwritten as well. 
In this case it is on purpose because we're removing the option for the parameter being optional.

## Expected behavior

No error / warnings from Pylance.

## Actual behavior

reportGeneralTypeIssues error.

**Comment by erictraut** [comment]
The semantics for pydantic appear to be slightly different from the stdlib dataclass in this case. Pyright's behavior and error message is correct for a stdlib dataclass. Pydantic deviates from this behavior, and standards-based type checkers like pyright do not know about pydantic's specific idiosyncrasies. With a regular dataclass, you can see that pyright's error here makes more sense.

```python
from dataclasses import dataclass

@dataclass
class A:
    a: int | None = None

@dataclass
class B(A):
    a: int

b = B()
print(b.a) # Prints "None"
```

In any case, pyright is working correctly here and in conformance with the Python typing spec. If you'd like to report this issue to the maintainers of pydantic, you're welcome to do so.

**Comment by ViktorClaesson** [comment]
Okay I understand, fair enough, thanks a lot for the explanation and for taking your time!
